[SPMD] Introduce the cross-iteration graph optimization framework

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94803

Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

fegin committed:
[SPMD] Introduce the cross-iteration graph optimization framework

Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

anj-s commented on torch/distributed/_spmd/iter_graph_module.py:
```diff
@@ -0,0 +1,393 @@
+        return self._insert_context("inserting_before", node)
+
+    @staticmethod
+    def _is_sese_graph(nodes: List[fx.Node], graph: fx.Graph) -> bool:
```
Does this mean that the output of this graph cannot be the input of two following subgraphs?

fegin replied:
It can but there can only be one output. However, I found that this is to restrictive. Will change this in the next few PRs.

anj-s commented on torch/distributed/_spmd/iter_graph_module.py:
```diff
@@ -0,0 +1,393 @@
+        # new_output =  tuple(
+        #   arg for arg in output.args if arg != nodes[-1]
+        # ) + (arguments,)
+        new_output = output.args + (sese_arguments,)
```
Isn't the first_node the input node to this graph? Can you clarify why the input and output of the graph is converted to be the new output? I would expect only the output to be converted to input.

fegin replied:
Since the entire subgraph will be moved, both the input and output of the subgraph will be modified. The input will from the `last iteration` which will be passed in through the input of the full graph. The output will become the input of the `next iteration` which will be carried over via the output of the full graph.

anj-s commented on torch/distributed/_spmd/iter_graph_module.py:
```diff
@@ -0,0 +1,393 @@
+        #   arg for arg in output.args if arg != nodes[-1]
+        # ) + (arguments,)
+        new_output = output.args + (sese_arguments,)
+        if erase_node:
```
Why do we erase all the graph nodes?

anj-s commented on torch/distributed/_spmd/iter_graph_module.py:
```diff
@@ -0,0 +1,393 @@
+                    output = output[0]
+            return output
+        else:
+            return gm(*args, **kwargs)
```
why do we not consider self._previous_output in the args for this FW?

fegin replied:
For this case, this mean all the optimizations do not produce any cross-iteration movement. As a result, no extra/previous output exist. We have to do this as IterGraphModule will be the entry point of all the optimizations of all the graphs but not all graphs may be required to do cross-iteration movement (if algorithm believes no benefit).

anj-s commented on torch/distributed/_spmd/iter_graph_module.py:
```diff
@@ -0,0 +1,393 @@
+    def print_all_gms(self) -> None:
+        logger.info("Printing the three fx gm:")
+        logger.info("1. Setup gm:")
+        logger.info("{self.setup_gm.print_readable(False)}")
```
missing "f" for correct formatting

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce the cross-iteration graph optimization framework"


Introduce the cross-iteration graph optimization framework that allow users to write a graph optimization that moves nodes cross iterations.

Differential Revision: [D43247247](https://our.internmc.facebook.com/intern/diff/D43247247/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
